### ISSUE TITLE
feat: pagination support

### ISSUE THREAD

**Issue by kvithana** [ndench/up-bank-api#5]
Should we try to support pagination for list endpoints? I'm not sure what the best practice would be for support this. 

What do you think of something like this which wraps around the response objects which allow pagination, which replaces the links with the api client method to get more data (keeping the interface type).
```typescript
interface PaginationMethods {
  links: {
    next: () => Promise<unknown> | null;
    prev: () => Promise<unknown> | null;
  };
}

export function withPagination<T>(
  client: UpClient,
  data: T
): T & PaginationMethods {
  const links: PaginationLinks = data.links;
  return {
    ...data,
    links: {
      next: links.next
        ? () =>
            client
              .get<T>(links.next.replace(BASE_URL, ''))
              .then((data) => withPagination(client, data))
        : null,
      prev: links.prev
        ? () =>
            client
              .get<T>(links.prev.replace(BASE_URL, ''))
              .then((data) => withPagination(client, data))
        : null,
    },
  };
}
```

Then for methods that support pagination we could use this builder link so:
```typescript
  return this.api
      .get<ListAccountResponse>(`${ENDPOINT}?${urlParams.join('&')}`)
      .then((data) => withPagination(this.api, data));
```

**Comment by ndench** [comment]
This is a really cool idea @kvithana, I hadn't even considered it.

Is your intention that the pagination could be used something like this:

```typescript
let accounts = await up.accounts.list({ pageSize: 30 });
doSomethingWithAccounts(accounts);

while (accounts.next) {
    accounts = await accounts.next();
    doSomethingWithAccounts(accounts);
}
```

Do you have any other example SDKs that do pagination (either like this or in another way)? Doesn't have to be TypeScript SDKs (although that would be preferable), it would be good to see if there are other things we need to consider or issues other people have solved before.

**Comment by kvithana** [comment]
Hmmm it's not something I've run into before. The last API wrapper library I used just ignored the prev/next links in the response (like how we're doing currently) and have required the user to explicitly specify the offset & limit on each request. ie.
```typescript
// request 1
api.getMyThings(uid, { limit : 20 })
// request 2
api.getMyThings(uid, { limit: 20 , offset : 20 })
```
I'll try to have a look around.

But yes my intention was to be able to user it as an iterator, since all it does it wrap the next/prev URLs with the existing authentication headers and return a function which will make the GET request. But note in your example, it would be `accounts.links.next()` if we were to follow structure of the original response object.

**Comment by ndench** [comment]
Yeah fair enough, that sounds cool. If you want to have a crack at implementing it, go for it.

Make sure to update the readme with some examples of how it should be used in an iterator.

**Comment by ndench** [comment]
Hey @kvithana how are you coming along with this feature? Do you need any help?

**Comment by kvithana** [comment]
Hi @ndench . I've been sitting on this for a while because I'm not sure what the best implementation approach would be.

I currently have implemented all paginable endpoints to return an async iterator. This means that you **must** iterate over the items to extract the amount of data that you need. Some people are comfortable working with iterators in this way and it allows us to abstract away the pagination entirely. However it does require a change in usage pattern. As an example, here's how you would use the list accounts endpoint:

```typescript
// gets iterator, accounts variable does not contain any data
const accounts = up.accounts.list();

// can iterate over it with pagination abstracted
for await (const account of accounts) {
  console.log(account);
}

// utility function to iterate over all
await takeAll(accounts);
// utility function to get n items
await take(accounts, 5);
```

The benefit of this approach is that paginated endpoints return a generic javascript iterable, compatible with all types of async iterable libraries which let you do powerful map/reduce/filter operations on the iterable. However, as you can see this implementation doesn't give you data directly by calling the `.list()` method. 

Some additional work _could_ be done to support both these use patterns, perhaps by supporting something like `up.accounts.list().take(50)` or `up.accounts.list.takeAll()` but some refactoring and thought would need to go into how to set this up elegantly. 

Thoughts?

**Comment by ndench** [comment]
I love the idea of returning an iterator! It would be awesome to be able to use map/reduce/filter/etc. And having something like `take(50)` and `.takeAll()` would be great as well.

There are 2 things I don't love about it:
1. As you mentioned, you **must** use the iterator. I think there are definitely times when it would be better to not use an iterator, and I don't want to force people to always use iterators.
2. We would be diverging from the official Up api docs, in that `accounts.list` doesn't return what the docs say.

However, I think we can easily solve these issues by leaving the `up.accounts.list()` function as it is, and making a new `up.accounts.iterator()` or `up.accounts.iterate()` function. I think this is a better name for it and makes it more obvious what's happening as well.

What do you think?

I would still like original idea of replacing the `.next` and `.previous` properties with functions as well. But I think this iterator idea is must better.

**Comment by IAmMadfly** [comment]
I gave it a crack to change the incoming next/prev object values from string | null to () => Promise<T> | null.

Let me know what you think of my implementation. I did it at the point where we receive the information. This allows for any type that may have pagination to have next() and prev() functions, removing the need to implement it for every individual type.
